executor/conn.h

```diff
@@ -209,7 +209,10 @@ class Select
 	void Wait(int ms)
 	{
 		timespec timeout = {.tv_sec = ms / 1000, .tv_nsec = (ms % 1000) * 1000 * 1000};
-		if (pselect(max_fd_ + 1, &rdset_, nullptr, nullptr, &timeout, nullptr) < 0) {
+		for (;;) {
+			if (pselect(max_fd_ + 1, &rdset_, nullptr, nullptr, &timeout, nullptr) >= 0)
```

Comment:
There is something fishy going on here.
1. Ready function checks that the descriptor was armed, so we shouldn't read from non-ready sockets.
2. All file descriptors should be in non-blocking mode, so even if we were wrong with signalling, it still should not hand.

Is it another bug in Starnix?

Additionally:
3. We actually do want to unblock on SIGCHLD, that's very intentional, we want to wait for the child and restart it.
4. We still can return with 0 ready sockets on the timeout, we check `>= 0` and 0 means no descriptors ready.

Reply:
eepeep@ will be back in a couple of days, so this might need to wait for their knowledgeable eye. I've spent some time today trying to understand the root cause, but it's still not clear to me.

I think I agree with your comments, and I've checked that non-blocking mode is implemented for pipes in Starnix. It looks like read() is behaving correctly and returning EAGAIN when the write end of the pipe is open and there is no data to read.

A hack that fixes the hang is a 1ms sleep just before this line: https://github.com/google/syzkaller/blob/6c853ff934ae691d82d2ddf8d401fdd07ed4ab74/executor/executor_runner.h#L537 With that hack, I see several `syz-executor exec` processes starting (and restarting with different PIDs, eventually) during the machine check stage. Without it, several `syz-executor exec` processes run briefly during the machine check, but die and are not replaced. The executor runner might be waiting for data from one of these missing exec processes during the hang.

I'm not sure yet if eepeep's retry fix here is directly addressing the root cause, or if it's introducing a different tiny timing change that has the same effect as the sleep hack.

Any advice would be welcome, otherwise I'll keep looking at this tomorrow.

Reply:
I don't have any advice regarding the root cause because since it's Starnix, it's too easy to write everything off to some difference in behavior. But I don't know if this difference really exists and if it's a legit difference or not.

Regarding debugging, I would first try to understand actual state of the process (and subprocesses) when it hangs. Adding more printf's looks like the simplest first step. It may also be possible to start a background thread that will sleep for some time and then dump state of everything.

Reply:
1. `Ready()` checks `FD_ISSET(fd, &rdset_)` but this is the same `rdset_` passed as input to `pselect`, and `pselect` does not clear its `readfds` argument when returning an error like this[1].  That means the `Ready` check trivially and incorrectly passes because it sees the appropriate FD bit still as it was set by `Arm` earlier.

[1] From `man pselect`:
>  On  error,  -1  is returned, and errno is set to indicate the error; the
>    file descriptor sets are unmodified, and timeout becomes undefined.

2. It does not block on the socket read calls, but `Recv` has an [outer loop](https://github.com/google/syzkaller/blob/ee2602b8aa3e172d05dbaff914a78720b3dd04f9/executor/conn.h#L81) that loops infinitely on `read()`s that keep returning `EAGAIN`.

In my understanding, this is all expected behavior and not an issue with Starnix. Does this clarify the confusion?

Reply:
Also, to clarify, the reads here are on the connection with syz-manager, not a subprocess.

Reply:
You are right.